---
title: "R more"
subtitle: "Best practices using R in 2018"
output: 
    html_document:
        theme: cerulean
        css: style.css
---

<br>
These are the course notes for the [Monash Bioinformatics Platform](https://www.monash.edu/researchinfrastructure/bioinformatics)'s "R More" course.
This builds on our ["Introduction to R"](https://monashdatafluency.github.io/r-intro-2/) course. These modules cover best practices for reproducable sharable results, programming, "tidy" exploratory data analysis, and presenting data interactively with Shiny. We will be covering much more ground than the introductory course, to give a broad overview of the various ways R is currently used.

There is a further module on manipulating DNA sequences and features with Bioconductor. This turned out to be a bit much for one day, and is also more specialized than other topics, so will generally be taught separately.

## Files

* [r-more-files.zip](r-more-files.zip) - files used in this course

## Slides

* [Introduction](slides/introduction.html) ([.Rmd](slides/introduction.Rmd))
* [Shiny](https://bioinformatics3.erc.monash.edu/rsconnect/content/263/#1) ([.Rmd](slides/shiny_course_notes.Rmd))

## Topics

* [Rmarkdown](topics/Rmarkdown.html) ([.Rmd](topics/Rmarkdown.Rmd))
* [Programming](topics/programming.html) ([.Rmd](topics/programming.Rmd))
* [The tidyverse: dplyr, ggplot2, and friends](topics/tidyverse.html) ([.Rmd](topics/tidyverse.Rmd))
* [Shiny](https://bioinformatics3.erc.monash.edu/rsconnect/content/79/) ([.Rmd](topics/shiny_tutorial.Rmd))

## Further topics

* [Sequences and genomic features](topics/sequences_and_features.html) ([.Rmd](topics/sequences_and_features.Rmd))

---

## See also

In this series:

* [Introduction to R (version 2)](https://monashdatafluency.github.io/r-intro-2/)
* [RNAseq DE analysis using R](http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/)

Other resources:

* [Next steps](https://monashdatafluency.github.io/r-intro-2/next-steps.html)


## Authors

This course is developed by the Monash Bioinformatics Platform:

* Paul Harrison
* Adele Barugahare 
* Kirill Tsyganov

<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="figures/CC-BY.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

Some of this material is derived from work by [Software Carpentry](http://software-carpentry.org/). Any errors and failures of pedagogy are our own.

## Source code

* [GitHub repository](https://github.com/MonashBioinformaticsPlatform/r-more)




